tests/__init__.py

```diff
@@ -0,0 +1,11 @@
+# This file is necessary to be able to launch pytest by running "pytest"
+
+import sys
+from pathlib import Path
+
+# Get the sources folder absolute path
+project_path = Path(__file__).parent.parent
+sources_path = project_path.joinpath("src")
+
+# Necessary for pytest to allow imports from packaged source folders
+sys.path.append(str(sources_path))
```

Comment:
Is this true that it is really necessary? `pyhf`'s `tests/__init__.py` is blank and we're able to run without this.

Reply:
This is necessary if we want to import specific classes / functions / ... directly from a module within the source folder.

Steps to reproduce:
1. Define a dummy class within a new `src/pyhf_extractor` module.
2. Define a dummy test within a new `test` module.
3. Import that class in the test doing `from pyhf_extractor.<module_name> import <dummy_class>`.
4. Run `pytest` (or `make test`) from the root folder.

You will see it run successfully.
Now, if you remove the referenced lines, that type of imports will not work.

Reply:
just update your `setup.cfg`

```
[options]
setup_requires =
    setuptools_scm>=1.15.0
    setuptools_scm_git_archive>=1.0
package_dir =
    pyhf = src/pyhf
    pyhf_extractor = src/pyhf_extractor
```

Reply:
There is no `setup.cfg` yet.

I am not sure what `setup_scm_git_archive` does, but I am reluctant to define a random package hosted by a random guy, with ~10 stars, as a dependency.

Finally, what if there are multiple folders within `src`? Is the `package_dir` structure ready to support multiple ones?

Reply:
Yes. It's baked in python. `setup.cfg` is a one-to-one replacement for `setup.py`. The documentation I linked is from python's central docs.

Reply:
I think for getting this in and moving forward it doesn't matter too much, as it is easy to change later, but for reference if it is helpful this is what `pyhf` does for `setup.py` and `setup.cfg` which I'd like to think makes things quite minimal and readable:
- [`setup.py`](https://github.com/scikit-hep/pyhf/blob/master/setup.py)
- [`setup.cfg`](https://github.com/scikit-hep/pyhf/blob/master/setup.cfg)

Reply:
Oh i see, so the `package_dir` structure maps with the one on the `setup` function call. I guess it is some sort of key-value "dictionary"...

In any case, same comment as above:

I am not seeing how `setup.py` is difficult to follow as it is right now. Sure, if it grows we can split it into 2 files, but I don't see any benefit with its current state.

Reply:
> I am not seeing how `setup.py` is difficult to follow as it is right now

It isn't difficult, but is also largely unnecessary.

> Sure, if it grows we can split it into 2 files, but I don't see any benefit with its current state.

I think the basic argument is that this is the way that the PyPA suggests people develop libraries: `setup.py` is for logic as it is Python, and `setup.cfg` is for static configuration options. c.f. the [PyPA's page on "Packaging and distributing projects"](https://packaging.python.org/guides/distributing-packages-using-setuptools/#configuring-your-project) for a brief discussion.

This is not a discussion that should hold up anything though, so I'd prefer to move forward with what you have and discuss this later if people care.